File: packages/feedback/src/ports/clock.ts
// M9 clock port (clean-room).
export interface Clock {
  now(): number;
}

export class SystemClock implements Clock {
  now(): number {
    return Date.now();
  }
}

export class FakeClock implements Clock {
  private value: number;
  constructor(start = 0) { this.value = start; }
  now(): number { return this.value; }
  advance(ms: number): void { this.value += ms; }
}

File: services/ai-worker/src/fastwork_ai_worker/handoff/clock.py
"""Injected clock (M9, clean-room). Business logic never calls datetime.now directly."""
from __future__ import annotations

import time
from typing import Protocol


class Clock(Protocol):
    def now_ms(self) -> int: ...


class SystemClock:
    def now_ms(self) -> int:
        return int(time.time() * 1000)


class FakeClock:
    def __init__(self, start_ms: int = 0):
        self._now = start_ms

    def now_ms(self) -> int:
        return self._now

    def advance(self, ms: int) -> None:
        self._now += ms
